src/rfft!.jl

```diff
@@ -0,0 +1,234 @@
+import Base: IndexStyle, getindex, setindex!, eltype, \, similar, copy, real, read!
+
+export PaddedRFFTArray, plan_rfft!, rfft!, plan_irfft!, plan_brfft!, brfft!, irfft!
+
+
+
+# As the time this code was written the new `ReinterpretArray` introduced in
+# Julia v0.7 had major performace issues. Those issues were bypassed with the usage of the
+# custom getindex and setindex! below. Hopefully, once the performance issues with ReinterpretArray
+# are solved we can just index the reinterpret array directly.
+
+struct PaddedRFFTArray{T<:fftwReal,N,L} <: DenseArray{Complex{T},N}
+    data::Array{T,N}
+    r::SubArray{T,N,Array{T,N},NTuple{N,UnitRange{Int}},L} # Real view skipping padding
+    c::Base.ReinterpretArray{Complex{T},N,T,Array{T,N}}
+
+    function PaddedRFFTArray{T,N}(rr::Array{T,N},nx::Int) where {T<:fftwReal,N}
+        rrsize = size(rr)
+        fsize = rrsize[1]
+        iseven(fsize) || throw(
+            ArgumentError("First dimension of allocated array must have even number of elements"))
+        (nx == fsize-2 || nx == fsize-1) || throw(
+            ArgumentError("Number of elements on the first dimension of array must be either 1 or 2 less than the number of elements on the first dimension of the allocated array"))
+        fsize = fsize÷2
+        csize = (fsize, rrsize[2:end]...)
+        c = reinterpret(Complex{T}, rr)
+        rsize = (nx,rrsize[2:end]...)
+        r = view(rr,(1:l for l in rsize)...)
+        return  new{T, N, N === 1 ? true : false}(rr,r,c)
+    end # function
+end # struct
+
+@inline real(S::PaddedRFFTArray) = S.r
+
+@inline complex_view(S::PaddedRFFTArray) = S.c
+
+@inline data(S::PaddedRFFTArray) = S.data
+
+copy(S::PaddedRFFTArray) = PaddedRFFTArray(copy(data(S)),size(real(S),1))
+
+similar(f::PaddedRFFTArray,::Type{T},dims::Tuple{Vararg{Int,N}}) where {T, N} =
+    PaddedRFFTArray{T}(dims) 
+similar(f::PaddedRFFTArray{T,N,L},dims::NTuple{N2,Int}) where {T,N,L,N2} =
+    PaddedRFFTArray{T}(dims) 
+similar(f::PaddedRFFTArray,::Type{T}) where {T} =
+    PaddedRFFTArray{T}(size(real(f))) 
+similar(f::PaddedRFFTArray{T,N}) where {T,N} = 
+    PaddedRFFTArray{T,N}(similar(data(f)), size(real(f),1)) 
+
+size(S::PaddedRFFTArray) =
+    size(complex_view(S))
+
+IndexStyle(::Type{T}) where {T<:PaddedRFFTArray} = 
+    IndexLinear()
+
+@inline function getindex(A::PaddedRFFTArray{T,N}, i2::Integer) where {T,N}
+    d = data(A)
+    i = 2i2
+    @boundscheck checkbounds(d,i)
+    @inbounds begin 
+        return Complex{T}(d[i-1],d[i])
+    end
+end    
+
+@inline @generated function getindex(A::PaddedRFFTArray{T,N}, I2::Vararg{Integer,N}) where {T,N}
+    ip = :(2*I2[1])
+    t = Expr(:tuple)
+    for i=2:N
+        push!(t.args,:(I2[$i]))
+    end
+    quote
+        d = data(A)
+        i = $ip
+        I = $t
+        @boundscheck checkbounds(d,i,I...)
```

Comment:
Shouldn't it also check bounds with `i-1`?

Reply:
Oh, I guess since `ip=2*I2[1]`, if it is in-bounds then `i-1` must also be in-bounds.